src/Windows/dll/local_cache.cpp

```diff
@@ -1,33 +1,412 @@
+            if (fileName.compare(L".") != 0 && fileName.compare(L"..") != 0)
+            {
+                std::wstring fullFileName = baseDir + L"\\" + fileName;
+
+                throw_if(!DeleteFileW(fullFileName.c_str()),
+                    "Deleting file failed, error code " + GetLastError());
+            }
+        } while (FindNextFile(hFind, &data));
+        FindClose(hFind);
+    }
+
+    return;
+};
+
+HANDLE OpenHandle(LPCWSTR lpFileName, DWORD dwDesiredAccess, DWORD dwShareMode, LPSECURITY_ATTRIBUTES lpSecurityAttributes,
+    DWORD dwCreationDisposition, DWORD dwFlagsAndAttributes, HANDLE hTemplateFile) {
+    
+    HANDLE file;
+    int i = 0;
+    do {
+        file = CreateFile(lpFileName, dwDesiredAccess, dwShareMode, lpSecurityAttributes,
+            dwCreationDisposition, dwFlagsAndAttributes, hTemplateFile);
+        
+        Sleep(SLEEP_RETRY_MS);
+        i++;
+    } while ((file == INVALID_HANDLE_VALUE) &&
+        (GetLastError() == 32) &&
+        (i < MAX_RETRY));
+
+    return file;
+}
+
+
+void local_cache_add(const std::string& id, time_t expiry, size_t data_size, const void* data)
+{
+    throw_if(id.empty(), "The 'id' parameter must not be empty.");
+    throw_if(data_size == 0, "Data cannot be empty.");
+    throw_if(data == nullptr, "Data pointer must not be NULL.");
+
+    init();
+    CacheEntryHeaderV1 header{};
+    header.version = CACHE_V1;
+    header.expiry = expiry;
+
+    std::string filename = get_file_name(id);
+    std::wstring wfilename(filename.begin(), filename.end());
+
+    HANDLE file = OpenHandle(wfilename.c_str(), GENERIC_WRITE, 0, NULL, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, NULL);
+    throw_if(file == INVALID_HANDLE_VALUE, "Open handle failed");
+
+    OVERLAPPED overlapvar = { 0 };
+
+    throw_if(!LockFileEx(file, LOCKFILE_EXCLUSIVE_LOCK, 0, MAXDWORD, MAXDWORD, &overlapvar),
+        "Exclusive write lock operation failed");
+
+    DWORD headerwritten;
+    DWORD datawritten;
+
+    throw_if(!WriteFile(file, &header, sizeof(header), &headerwritten, NULL),
+        "Header write to local cache failed");
+
+    throw_if(!WriteFile(file, data, (DWORD)data_size, &datawritten, NULL),
+        "Data write to local cache failed");
+
+    throw_if(!UnlockFileEx(file, 0, MAXDWORD, MAXDWORD, &overlapvar),
+        "Exclusive write unlock operation failed");
+
+    CloseHandle(file);
 }
 
 std::unique_ptr<std::vector<uint8_t>> local_cache_get(
     const std::string& id)
 {
-    (void)id;
-    return nullptr;
+    throw_if(id.empty(), "The 'id' parameter must not be empty.");
+    init();
+
+    std::string filename = get_file_name(id);
+    std::wstring wfilename(filename.begin(), filename.end());
+    
+    HANDLE file = OpenHandle(wfilename.c_str(), GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+    if (file == INVALID_HANDLE_VALUE)
+    {
+        return nullptr;
+    }
+
+    OVERLAPPED overlapvar = { 0 };
```

Comment:
Since the file wasn't opened for async I/O, why do you have an OVERLAPPED here?

Reply:
This was definied in sample code online that I looked at, do you have another way you would suggest?

Reply:
Removed.